### ISSUE TITLE
SignomialEquality with trust regions

### ISSUE THREAD

**Issue by mayork** [convexengineering/gpkit#820]
I am trying to use a SignomialEqualityLinTrust and am getting the following error

```
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/gpkit/constraints/set.pyc in as_gpconstr(self, x0)
    181 
    182         When x0 is none, may return a default guess."""
--> 183         gpconstrs = [constr.as_gpconstr(x0) for constr in self]
    184         return ConstraintSet(gpconstrs,
    185                              self.substitutions, recursesubs=False)

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/gpkit/constraints/set.pyc in as_gpconstr(self, x0)
    181 
    182         When x0 is none, may return a default guess."""
--> 183         gpconstrs = [constr.as_gpconstr(x0) for constr in self]
    184         return ConstraintSet(gpconstrs,
    185                              self.substitutions, recursesubs=False)

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/gpkit/nomials/nomial_math.pyc in as_gpconstr(self, x0)
    888         varkeys = set(mleft.varkeys).union(mright.varkeys)
    889     bounding_constraints = [[self.trustregion <= Monomial(vk)/x0[vk],
--> 890                          Monomial(vk)/x0[vk] <= 1/self.trustregion] for vk in varkeys]
    891         return [bounding_constraints, mleft == mright]
    892 

```

I can run the example posted in #808 fine. This is  the method throwing the error

``` python
def as_gpconstr(self, x0):
        "Returns GP apprimxation of an SP constraint at x0"
        #return self.left >= self.right
        def _force_mono(posy):
            if isinstance(posy, Monomial):
                return posy
            return posy.mono_lower_bound(x0)
        mleft, mright = map(_force_mono, [self.left, self.right])
    varkeys = set(mleft.varkeys).union(mright.varkeys)
    bounding_constraints = [[self.trustregion <= Monomial(vk)/x0[vk],
                         Monomial(vk)/x0[vk] <= 1/self.trustregion] for vk in varkeys]
    return [bounding_constraints, mleft == mright]
```

I'm not 100% sure but I think the problem is that no argument is being passed into `_force_mono`

@mopg this might be of interest to you

**Comment by mayork** [comment]
throwing this error SignomialEqualityTrivTrust

```
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/gpkit/constraints/set.pyc in as_gpconstr(self, x0)
    181 
    182         When x0 is none, may return a default guess."""
--> 183         gpconstrs = [constr.as_gpconstr(x0) for constr in self]
    184         return ConstraintSet(gpconstrs,
    185                              self.substitutions, recursesubs=False)

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/gpkit/nomials/nomial_math.pyc in as_gpconstr(self, x0)
    781                 return posy
    782             return posy.mono_lower_bound(x0)
--> 783         return [_force_mono(self.left, x0) >= self.trustregion*self.right,
    784                 self.left <= _force_mono(self.right, x0)]
    785 

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/gpkit/nomials/nomial_math.pyc in _force_mono(posy, x0)
    780             if isinstance(posy, Monomial):
    781                 return posy
--> 782             return posy.mono_lower_bound(x0)
    783         return [_force_mono(self.left, x0) >= self.trustregion*self.right,
    784                 self.left <= _force_mono(self.right, x0)]

AttributeError: 'Signomial' object has no attribute 'mono_lower_bound'
```

less sure what is causing this...maybe I'm just using these wrong. I tried to just replace my `SignomialEqality(x,y)' with `SignomialEqualityTrvTrust(x,y)` and use the default trust region

**Comment by mopg** [comment]
Yeah, this method is not the robust method ever written. Just explicitly pass an initial guess (can be 1’s) for the variables appearing in the equality constraint, and then it should work.

**Comment by mopg** [comment]
For SignomialEqualityTrivTrust(), did you reorder the equality such that there is a posynomial on both sides? It should take care of that, but maybe that is going wrong

**Comment by mayork** [comment]
no I didn't reorder, just tried to plug and play. 

As far as passing in an initial guess is that what's occurring with the `x0={x.key:1,y.key:1}` in 
```
sol = m.localsolve(verbosity=2, x0={x.key:1,y.key:1}, iteration_limit=101)
```

**Comment by mayork** [comment]
also, is the old 'SignomialEquality' 'SignomialEqualityLin' or `SignomialEqualityTriv`

**Comment by mopg** [comment]
Yes that is passing the initial guess.

**Comment by mopg** [comment]
SignomialEquality on master is SignomialEqualityLin on equality

**Comment by mayork** [comment]
manually setting the initial guess did fix the problem with 'SignomialEqualityLinTrust'

**Comment by mayork** [comment]
I don't see a pull request for this branch, am I correct there is not one yet? If there isn't, is it okay if I make one? I'm getting some weird infeasible constraint errors that I don't believe so I'm curious to see if this passes unit tests

**Comment by mayork** [comment]
so i can avoid the ``Attribute Error`` in ``SignomialEqualityLin`` by making all my constraints of the form posynomial=posynomal. Most of the signomial equality constraints I have are ``p=f+1`` so I changed these to be ``p+1=f+2``. When I do this I get this error

```
ValueError: infeasible constraint: 1 >= 2
```

which traces back to line 830 of nomial_math.py. That line of code is 

```python
return [_force_mono(self.left, x0) >= _force_mono(self.right, x0),
```

inside the method as_gpconstr in the class ``SignomialEqualityLin``.

If I got back to my model I run on master and change all my p=f+1 constraints to be p+1=f+2 everything runs just fine. So I think there's an error with the _force_mono call. Not sure what it is but I'll take a look...although I think someone else is probably better suited to figure out exactly what's going on

**Comment by bqpd** [comment]
The version of SignomialEquality on master doesn't use `_force_mono`, since you're now allowed to call monomial_approx on a monomial. Merging master and not using `_force_mono` anymore might fix this.

**Comment by whoburg** [comment]
@mayork, try now.

@mopg already merged master into equality, and I just removed `_force_mono`.

**Comment by whoburg** [comment]
... although I'm not sure the removal of `force_mono` will fix the `'Signomial' object has no attribute 'mono_lower_bound'`. @bqpd, @mopg, I think maybe `self.left`, `self.right` should be replaced with posy, negy?

**Comment by mayork** [comment]
just tried this branch as suspected the ``AttributeError: 'Signomial' object has no attribute 'mono_lower_bound'`` is still thrown

**Comment by bqpd** [comment]
Oh, yes, that too. I did that in master and forgot to backport.

**Comment by whoburg** [comment]
#835 should make this easier to fix - once reviewed we can make the changes on `equality`

**Comment by mayork** [comment]
I saw #835 was merged in, have those changes been taken into account on ``equality``? I think trust regions must be used in the fitted atmosphere model I'm trying to make

**Comment by bqpd** [comment]
I should merge equality onto master, I think, if we're going to do active development on it.

**Comment by bqpd** [comment]
@mayork, check out the `sigtrust` branch and see if it does what you need.

**Comment by bqpd** [comment]
Superseded by #869

**Comment by bqpd** [comment]
Reopened

**Comment by mayork** [comment]
@1ozturkbe

**Comment by whoburg** [comment]
@mayork, are you sure about the title change? Is this issue about the error being raised or about a model that doesn't terminate without a trust region? If the latter, we need the Model, a MWE, etc.

**Comment by mayork** [comment]
oh shoot I think I changed the title of the wrong issue, I'll change it back. I actually think this issue can be closed since @bqpd has created the branch where he fixed the trust region implementation.

There is an issue where w/no trust a region a model bounces between a solution that's not feasible and a feasible solution. I'll go look for that one.

**Comment by mayork** [comment]
yeah I was thinking of 808, adjusted that title accordingly. It has the commit number where the problem occurred. I think the long term plan was @1ozturkbe was going to take a look at it and see if we can't figure out what's going on

**Comment by whoburg** [comment]
great, yes that will be interesting to understand better.

BTW, re: closing this, general guidance is to not close the issue until the fix is merged to master. That's why @bqpd created the "waiting on PR" label.

**Comment by bqpd** [comment]
Are these still desired?

**Comment by mayork** [comment]
i think so

**Comment by bqpd** [comment]
Great!
